### ISSUE TITLE
[Bug Report][alpha branch]JumpReLU SAEs not reducing L0 in training

### ISSUE THREAD

**Issue by evan-lloyd** [jbloomAus/SAELens#494]
**Describe the bug**
On the current alpha branch, JumpRELU SAEs fail to reduce L0 in training, regardless of L0 coefficient. See for example this wandb plot:

![Image](https://github.com/user-attachments/assets/e8ce70cb-2bec-4045-98d8-04618164cd08)

**Minimal example/context**
Context: I'm experimenting with a novel SAE architecture, implemented as a subclass of `TrainingSAE`, and wanted to generate a baseline comparison using JumpReLU. When I swap the JumpReLU with a StandardSAE in the below setup, I get a reasonably sparse result, so this seems to be JumpReLU-specific.

Adapted from the [tutorial](https://github.com/jbloomAus/SAELens/blob/main/tutorials/training_a_sparse_autoencoder.ipynb), accounting for the refactors in alpha:

```
from sae_lens import (
    JumpReLUTrainingSAEConfig,
    LanguageModelSAERunnerConfig,
    SAETrainingRunner,
    TrainingSAE,
)
from sae_lens.config import LoggingConfig
from sae_lens.saes.sae import SAEMetadata

context_size = 1024
total_training_steps = 2_000
batch_size = 10 * context_size
total_training_tokens = total_training_steps * batch_size
model_name = "gpt2-small"
hook_layer = 5
hook_name = f"blocks.{hook_layer}.hook_resid_post"
dataset_path = "monology/pile-uncopyrighted"
expansion_factor = 4

cfg = LanguageModelSAERunnerConfig(
    eval_batch_size_prompts=16,
    model_name=model_name,
    hook_name=hook_name,
    hook_layer=5,
    sae=JumpReLUTrainingSAEConfig(
        metadata=SAEMetadata(
            model_name=model_name,
            hook_name=hook_name,
            hook_layer=hook_layer,
            context_size=context_size,
            dataset_path=dataset_path,
        ),
        d_sae=768 * expansion_factor,
        normalize_activations="expected_average_only_in",
        d_in=768,
        mse_loss_normalization="dense_batch",
        device="cuda",
        l0_coefficient=100,
        l0_warm_up_steps=total_training_steps // 20,  # 5% of training,
    ),
    dataset_path=dataset_path,
    is_dataset_tokenized=False,
    streaming=True,  # we could pre-download the token dataset if it was small.
    lr=1e-5,  # lower the better, we'll go fairly high to speed up the tutorial.
    adam_beta1=0.9,  # adam params (default, but once upon a time we experimented with these.)
    adam_beta2=0.999,
    lr_scheduler_name="constant",  # constant learning rate with warmup. Could be better schedules out there.
    lr_warm_up_steps=0,  # this can help avoid too many dead features initially.
    lr_decay_steps=total_training_steps // 5,  # 20% of training
    train_batch_size_tokens=batch_size,
    context_size=context_size,  # will control the lenght of the prompts we feed to the model. Larger is better but slower. so for the tutorial we'll use a short one.
    # Activation Store Parameters
    n_batches_in_buffer=64,  # controls how many activations we store / shuffle.
    training_tokens=total_training_tokens,  # 100 million tokens is quite a few, but we want to see good stats. Get a coffee, come back.
    store_batch_size_prompts=32,
    # Resampling protocol
    feature_sampling_window=1000,  # this controls our reporting of feature sparsity stats
    dead_feature_window=1000,  # would effect resampling or ghost grads if we were using it.
    dead_feature_threshold=1e-4,  # would effect resampling or ghost grads if we were using it.
    # WANDB
    logger=LoggingConfig(
        log_to_wandb=True,  # always use wandb unless you are just testing code.
        wandb_project="jumprelu-repro",
        wandb_log_frequency=30,
        eval_every_n_wandb_logs=20,
    ),
    # Misc
    device="cuda",
    seed=42,
    n_checkpoints=0,
    checkpoint_path="/tmp/checkpoints",
    dtype="float32",
)
# Looks like a bug in the alpha branch; the default code path filters out any config
# values not present in the base config class
override_sae = TrainingSAE.from_dict(cfg.get_training_sae_cfg_dict())
runner = SAETrainingRunner(cfg, override_sae=override_sae)

sae = runner.run()
```

Note also: I am using `override_sae` because [this code path](https://github.com/jbloomAus/SAELens/blob/f1c0586cd98efb3c8c7904ef1dc14de5f3343195/sae_lens/sae_training_runner.py#L82-L86) causes architecture-specific config options to be reset to default, due to instantiating a generic config class that then filters unknown options; I suspect this line should just be `TrainingSAE.from_dict(self.cfg.get_training_sae_cfg_dict())`.

**System Info**
Describe the characteristic of your environment:
 * Describe how `transformer_lens` was installed (pip, docker, source, ...)
 `uv` lockfile
 * What OS are you using? (Linux, MacOS, Windows)
 Linux
 * Python version (We suppourt 3.10 -3.12 currently)
3.11.5

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/jbloomAus/SAELens/issues) in the repo (**required**)

**Comment by hijohnnylin** [comment]
Sorry if this was already asked elsewhere and this may be a bad question: Does this issue happen on the non-alpha branch?

**Comment by evan-lloyd** [comment]
> Sorry if this was already asked elsewhere and this may be a bad question: Does this issue happen on the non-alpha branch?

Reasonable question! I hadn't tried on an official release since StandardSAE is good enough for my baseline purposes, but since I have this example already I just tried a run on `5.10.7` and I do get similar, possibly-buggy results where the L0 doesn't go down even after the warmup:

![Image](https://github.com/user-attachments/assets/e47c9ef4-056b-438b-b799-81eab38f20d8)

This was generated with

```
from sae_lens import LanguageModelSAERunnerConfig, SAETrainingRunner

context_size = 1024
total_training_steps = 2_000
batch_size = 10 * context_size
total_training_tokens = total_training_steps * batch_size
model_name = "gpt2-small"
hook_layer = 5
hook_name = f"blocks.{hook_layer}.hook_resid_post"
dataset_path = "monology/pile-uncopyrighted"
expansion_factor = 4

cfg = LanguageModelSAERunnerConfig(
    architecture="jumprelu",
    l1_coefficient=100,
    l1_warm_up_steps=total_training_steps // 20,  # 5% of training,
    eval_batch_size_prompts=16,
    model_name=model_name,
    hook_name=hook_name,
    hook_layer=5,
    mse_loss_normalization="dense_batch",
    normalize_activations="expected_average_only_in",
    expansion_factor=expansion_factor,
    d_in=768,
    dataset_path=dataset_path,
    is_dataset_tokenized=False,
    streaming=True,  # we could pre-download the token dataset if it was small.
    lr=1e-5,  # lower the better, we'll go fairly high to speed up the tutorial.
    adam_beta1=0.9,  # adam params (default, but once upon a time we experimented with these.)
    adam_beta2=0.999,
    lr_scheduler_name="constant",  # constant learning rate with warmup. Could be better schedules out there.
    lr_warm_up_steps=0,  # this can help avoid too many dead features initially.
    lr_decay_steps=total_training_steps // 5,  # 20% of training
    train_batch_size_tokens=batch_size,
    context_size=context_size,  # will control the lenght of the prompts we feed to the model. Larger is better but slower. so for the tutorial we'll use a short one.
    # Activation Store Parameters
    n_batches_in_buffer=64,  # controls how many activations we store / shuffle.
    training_tokens=total_training_tokens,  # 100 million tokens is quite a few, but we want to see good stats. Get a coffee, come back.
    store_batch_size_prompts=32,
    # Resampling protocol
    feature_sampling_window=1000,  # this controls our reporting of feature sparsity stats
    dead_feature_window=1000,  # would effect resampling or ghost grads if we were using it.
    dead_feature_threshold=1e-4,  # would effect resampling or ghost grads if we were using it.
    # WANDB
    log_to_wandb=True,  # always use wandb unless you are just testing code.
    wandb_project="jumprelu-repro",
    wandb_log_frequency=30,
    eval_every_n_wandb_logs=20,
    # Misc
    device="cuda",
    seed=42,
    n_checkpoints=0,
    checkpoint_path="/tmp/checkpoints",
    dtype="float32",
)
runner = SAETrainingRunner(cfg)

sae = runner.run()
```

**Comment by chanind** [comment]
I think this is may be an issue with how JumpReLU works and why it's so tricky to train. The default settings for the bandwidth we use from the original JumpReLU work is 0.001, which means that the threshold will get no gradient unless the activation happens to land with 0.001 of the current threshold. This happens very rarely, and thus the threshold doesn't change much. Also, if the threshold increases but is still low, you won't see any change to L0. If you increase the bandwidth parameter it should start reacting a lot earlier in training (e.g. try 0.01, or  0.1 or maybe even 1.0).

In the past I've found I needed to train for ~2B+ tokens to get JumpReLU to work. For instance, here's a pretty similar setup to what you shared at 2B tokens. The L0 is basically completely flat then finally starts decreasing at the end of training (I think it would keep going down if I let it keep training).

<img width="437" alt="Image" src="https://github.com/user-attachments/assets/74ddf556-8076-447f-a0c1-dd20720ef424" />

Notebook is here: https://colab.research.google.com/drive/1T-VyEFxNWlD0UDdMOU9lNjgvIFSMCExq?usp=sharing

Looking at [the paper](https://arxiv.org/pdf/2407.14435) again, it looks like they say the value of 0.001 is tuned assuming E[x] = 1, but we normalize to E[x] = sqrt(d_in) folling the Anthropic april update, so maybe we should default to a much higher bandwidth (e.g. 0.001 * sqrt(d_in)).

Anthropic also uses a much higher bandwidth parameter and a different kernel for their JumpReLU SAEs ([source](https://transformer-circuits.pub/2025/january-update/index.html)), which I would bet makes training a lot nicer (we still need to implement Anthropic's changes).

**Comment by hijohnnylin** [comment]
Thanks @chanind! This is mega useful. I've created this issue so that ideally we could have your explanation / example notebook equivalent in the repo: https://github.com/jbloomAus/SAELens/issues/498

**Comment by Hetens** [comment]
@chanind I tried the bandwidth explanation first, but I don't think it's the binding constraint. The gradient is reaching the threshold and the threshold still doesn't move.

Logging the fraction of pre-activations inside the STE window on current `main` with the v6 default `jumprelu_bandwidth=0.05` (gpt2 `blocks.5.hook_resid_post`, 2000 steps): **3.2% in-window on average**, i.e. ~100 of 3072 latents get threshold gradient every step. The threshold still went only `0.0100` → `0.0102`.

The blocker is the parameterization I think. Adam normalizes by gradient magnitude, so it moves a parameter by at most ~`lr` per step regardless of gradient size. With `threshold = exp(log_threshold)`, that's only ~`threshold * lr` of movement in real terms about `1e-7`/step at the `0.01` init. More gradient can't help, because Adam already normalized the magnitude away. Same reason `l0_coefficient` does nothing: it cancels between numerator and denominator.

Controlled runs (same seed/init/data, only the parameterization varies, bandwidth `0.05`, 2000 steps, `d_sae=3072`):

| | final threshold | final L0 |
|---|---|---|
| `exp(log_threshold)`, lr 3e-4 | 0.0157 | 3019 / 3072 |
| direct threshold, lr 3e-4 | **0.6197** | **1021** |

Runs: https://wandb.ai/hetansh-shah95-umass-amherst/saelens-494-jumprelu-threshold

Fix in #718 — parameterizes the threshold (without log) directly, as in DeepMind's pseudocode and `dictionary_learning`. Existing checkpoints are unaffected.

Two things worth checking would be:

- **Your 2B-token result may be the same effect.** The log parameterization makes the threshold move ~100x slower at a `0.01` init, which is roughly the factor by which training has to grow before anything shows. Untested at that scale, so just a hypothesis.
- **The fix doesn't make short runs sparse.** Adam still caps movement at ~`lr`/step, so reachable threshold ≈ `lr * steps`, and my runs sit near that ceiling. Tutorial-scale settings still end before L0 bends, so this probably wants lr/step guidance alongside it.

I'd be happy to add the bandwidth default into #718 if you'd like.

Here is a previous local run that i had

<img width="2250" height="630" alt="Image" src="https://github.com/user-attachments/assets/3b23af15-adea-4d8b-ab61-810148e01942" />
<img width="1650" height="630" alt="Image" src="https://github.com/user-attachments/assets/fdca778f-0bc7-4e9a-9164-733512f9a144" />
